### PR TITLE
Reinstate conditionals

### PR DESCRIPTION
**Pull request by cjllanwarne**
Including `simple_ifs`, `ifs_in_scatters` and `scatters_in_ifs`

### CODE REVIEW

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/stores/ExecutionStore.scala [hunk 1]
```diff
@@ -24,22 +24,20 @@ object ExecutionStore {
...
+        case _: ScatterGathererPort => None
+        case _ => key.index
       }
 
       key match {
         // For a scatter collector key, the dependencies are all the shards
-        case scatterCollector: ScatterCollectorKey => statusTable.row(scatterCollector.node).size == scatterCollector.scatterWidth
-        case conditionalCollector: ConditionalCollectorKey => statusTable.contains(conditionalCollector.node.executionNode, key.index)
+        case scatterCollector: ScatterCollectorKey =>
+          statusTable.row(scatterCollector.outputNodeToGather.singleInputPort.upstream.executionNode).size == scatterCollector.scatterWidth
```
Comment (Horneth) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/stores/ExecutionStore.scala]: I think dropping a comment here as to why we have to go find the `scatterCollector.outputNodeToGather.singleInputPort.upstream.executionNode` would be valuable, same below for conditional.

File: wom/src/main/scala/wom/values/WomOptionalValue.scala [hunk 2]
```diff
@@ -82,6 +82,28 @@ case class WomOptionalValue(innerType: WomType, value: Option[WomValue]) extends
...
+  def flattenToBaseType: WomOptionalValue = this match {
+    case WomOptionalValue(_: WomOptionalType, Some(innerOptionalValue: WomOptionalValue)) => innerOptionalValue.flattenToBaseType
+    case WomOptionalValue(innerType: WomOptionalType, None) => WomOptionalValue(innerType.baseMemberType, None)
+    case _ => this
+  }
+
+  private def boxUntilType(targetType: WomOptionalType): WomOptionalValue = {
+    assert(value.isDefined)
+    assert(targetType.baseMemberType.equals(womType.baseMemberType))
+    println(s"Boxing $womType to $targetType")
```
Comment (Horneth) [inline wom/src/main/scala/wom/values/WomOptionalValue.scala]: println

File: wom/src/main/scala/wom/values/WomOptionalValue.scala [hunk 3]
```diff
@@ -82,6 +82,28 @@ case class WomOptionalValue(innerType: WomType, value: Option[WomValue]) extends
...
+    if (womType.equals(targetType)) {
+      this
+    } else {
+      WomOptionalValue(womType, Some(this)).boxUntilType(targetType)
+    }
+  }
+
+  def coerceAndSetNestingLevel(womOptionalType: WomOptionalType) = this.flattenToBaseType match {
```
Comment (Horneth) [inline wom/src/main/scala/wom/values/WomOptionalValue.scala]: It's a bit hard to figure out what's happening, could you add a comment ?

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/keys/ScatterKey.scala [hunk 4]
```diff
@@ -59,14 +59,10 @@ private [execution] case class ScatterKey(node: ScatterNode) extends JobKey {
...
-    val collectors = node.outputMapping.groupBy(_.outputToGather.source.graphNode) flatMap {
-      case (taskCallNode: CallNode, scatterGatherPorts) => List(ScatterCollectorKey(taskCallNode, scatterGatherPorts, node, count))
-        // For other types of nodes we don't assume any time locality regarding the output ports availability and create
-        // one collector per port
-      case (declarationNode: AnonymousExpressionNode, scatterGatherPorts) => scatterGatherPorts map { sgp => ScatterCollectorKey(declarationNode, Set(sgp), node, count) }
-      case (_: ConditionalNode, scatterGatherPorts) => scatterGatherPorts map { sgp => ScatterCollectorKey(sgp.outputToGather, Set(sgp), node, count) }
-      case _ => List.empty
-    }
+    val collectors: Set[ScatterCollectorKey] = (node.outputMapping.groupBy(_.outputToGather.source.graphNode) flatMap {
+      case (_: CallNode | _: ExposedExpressionNode | _: ConditionalNode, scatterGatherPorts) => scatterGatherPorts.map(sgp => ScatterCollectorKey(sgp, count))
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/keys/ScatterKey.scala]: Why has the handling of anonymous vs exposed expression nodes changed?

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/stores/ExecutionStore.scala [hunk 5]
```diff
@@ -24,22 +24,21 @@ object ExecutionStore {
...
+        case _ => key.index
       }
 
       key match {
-        // For a scatter collector key, the dependencies are all the shards
-        case scatterCollector: ScatterCollectorKey => statusTable.row(scatterCollector.node).size == scatterCollector.scatterWidth
-        case conditionalCollector: ConditionalCollectorKey => statusTable.contains(conditionalCollector.node.executionNode, key.index)
+        case scatterCollector: ScatterCollectorKey =>
+          // The outputToGather is the PortBasedGraphOutputNode of the inner graph that we're collecting. Go one step upstream and then
+          // find the node which will have entries in the execution store. If that has 'n' entires, then we're good to start collecting,
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/stores/ExecutionStore.scala]: entries

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/stores/ExecutionStore.scala [hunk 6]
```diff
@@ -49,25 +48,22 @@ object ExecutionStore {
...
-    def executionNode: GraphNode = graphNode match {
-      case pbgon: PortBasedGraphOutputNode => pbgon.source.graphNode.executionNode
-      case other => other
-    }
 
-    def isInStatus(index: ExecutionIndex, table: StatusTable) = graphNode match {
-      // FIXME this won't hold conditionals in scatters for example
+    def isInStatus(index: ExecutionIndex, table: StatusTable): Boolean = graphNode match {
       // OuterGraphInputNode signals that an input comes from outside the graph.
       // Depending on whether or not this input is outside of a scatter graph will change the index which we need to look at
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/stores/ExecutionStore.scala]: this comment has been rudely separated from the code it references

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/keys/ScatterKey.scala (hunk 4 above)
Comment (cjllanwarne) [reply engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/keys/ScatterKey.scala]: This is just fixing a typo. Previously anonymous expressions (eg call input evaluations) were being collected and exposed outside the scatter and declarations were not.

File: centaur/src/main/resources/standardTestCases/conditionals_tests/scatters_in_ifs/scatters_in_ifs.wdl [hunk 7]
```diff
@@ -14,16 +14,17 @@ task mirror {
 
 workflow scatters_in_ifs {
 
-  Array[Int] numbers = range(3)
+# TODO: Reinstate the nested lookup with #2724
+#  Array[Int] numbers = range(3)
```
[the remark this answers is not in the document]
Comment (cjllanwarne) [reply centaur/src/main/resources/standardTestCases/conditionals_tests/scatters_in_ifs/scatters_in_ifs.wdl]: I'm already taking this as my next ticket!

**Comment by cjllanwarne** [comment]
@mcovarr anything particular you didn't get that you'd like to get?

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/keys/ScatterKey.scala (hunk 4 above)
Comment (mcovarr) [reply engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/keys/ScatterKey.scala]: oops

**Comment by mcovarr** [comment]
@cjllanwarne it can wait until the dust settles 🙂

**Pull request merged**